File: Backend/TrackrAPI/TrackrAPI/Services/Catalogo/DominioService.cs
using TrackrAPI.Dtos.Catalogo;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Catalogo;
using TrackrAPI.Helpers;

namespace TrackrAPI.Services.Catalogo
{
    public class DominioService
    {
        private readonly IDominioRepository dominioRepository;
        private readonly DominioValidatorService dominioValidatorService;

        public DominioService(
            IDominioRepository dominioRepository,
            DominioValidatorService dominioValidatorService)
        {
            this.dominioRepository = dominioRepository;
            this.dominioValidatorService = dominioValidatorService;
        }
        public DominioDto ConsultarDto(int idDominio)
        {
            var dominio = dominioRepository.ConsultarDto(idDominio);
            dominioValidatorService.ValidarExistencia(dominio);
            return dominio!;
        }

        public Dominio? Consultar(int idDominio)
        {
            return dominioRepository.Consultar(idDominio);
        }

        public Dominio? Consultar(string nombre)
        {
            return dominioRepository.Consultar(nombre);
        }

        public IEnumerable<DominioGridDto> ConsultarTodosParaGrid(int idUsuarioSesion)
        {
            return dominioRepository.ConsultarTodosParaGrid(idUsuarioSesion);
        }

        public IEnumerable<DominioDto> ConsultarTodosParaSelector()
        {
            return dominioRepository.ConsultarTodosParaSelector();
        }

        public int Agregar(Dominio dominio)
        {
            dominioValidatorService.ValidarAgregar(dominio);
            dominioRepository.Agregar(dominio);
            return dominio.IdDominio;
        }

        public void Editar(Dominio dominio)
        {
            dominioValidatorService.ValidarEditar(dominio);
            dominioRepository.Editar(dominio);
        }

        public void Eliminar(int idDominio)
        {
            var dominio = dominioRepository.Consultar(idDominio);

            if (dominio is null)
            {
                throw new CdisException("El dominio no existe");
            }

            dominioValidatorService.ValidarEliminar(idDominio);
            dominioRepository.Eliminar(dominio);
        }
    }
}

File: Frontend/trackr-administrador/src/app/shared/http/catalogo/dominio.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Dominio } from '@models/catalogo/dominio';

@Injectable()
export class DominioService {
  private dataUrl = 'dominio/';

  constructor(public http: HttpClient) {}

  public consultar(idDominio: number): Observable<Dominio> {
    return this.http.get<Dominio>(this.dataUrl + `consultar/${idDominio}`);
  }

  public consultarPorNombre(nombre: string): Observable<Dominio> {
    return this.http.get<Dominio>(
      this.dataUrl + `consultarPorNombre/${nombre}`
    );
  }

  public consultarTodosParaGrid(): Observable<Dominio[]> {
    return this.http.get<Dominio[]>(this.dataUrl + 'consultarTodosParaGrid');
  }

  public agregar(dominio: Dominio): Observable<number> {
    return this.http.post<number>(this.dataUrl + 'agregar', dominio);
  }

  public editar(dominio: Dominio): Observable<void> {
    return this.http.put<void>(this.dataUrl + 'editar', dominio);
  }

  public eliminar(idDominio: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idDominio}`);
  }

  public consultarGeneral() {
    return this.http.get<Dominio[]>(this.dataUrl + 'consultarGeneral');
  }

  public consultarTodosParaSelector() {
    return this.http.get<Dominio[]>(
      this.dataUrl + 'consultarTodosParaSelector'
    );
  }
}
